### PR TITLE
Fixed issue #18693: Using checkbox question for yes/no loses data when data entry view

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

File: application/controllers/admin/DataEntry.php [hunk 1]
```diff
@@ -601,30 +601,41 @@ public function editdata($subaction, $id, $surveyid)
...
-                    $aDataentryoutput .= " class='even'";
+                    $aDataentryoutput .= " class='even $questionTypeClass'";
                 }
 
                 $highlight = !$highlight;
-                $aDataentryoutput .= ">\n"
-                . "<td>"
-                . "\n";
+                $aDataentryoutput .= ">\n";
+                // First column (Question)
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:621]: :+1: we really need more information !

File: application/controllers/admin/DataEntry.php [hunk 2]
```diff
@@ -601,30 +601,41 @@ public function editdata($subaction, $id, $surveyid)
...
+                $aDataentryoutput .= "<td>";
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:622]: Maybe add a class here ? `question` or `question-cell` ?

File: application/controllers/admin/DataEntry.php [hunk 3]
```diff
@@ -601,30 +601,41 @@ public function editdata($subaction, $id, $surveyid)
...
                 $aDataentryoutput .= stripJavaScript($question);
-                $aDataentryoutput .= "</td>\n"
-                . "<td>\n";
+                $aDataentryoutput .= "</td>\n";
+                // Second column (Answer)
+                $aDataentryoutput .= "<td>\n";
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:626]: Maybe add a class here ? `answers` or `answers-wrapper`  or `answers-cell` ?

File: application/controllers/admin/DataEntry.php [hunk 4]
```diff
@@ -1325,6 +1329,23 @@ public function editdata($subaction, $id, $surveyid)
                         break;
                 }
 
+                if (!empty($questionInputs)) {
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:1332]: This was added for short-text and single choice question here ? (i must test quickly)

File: assets/packages/adminbasics/scss/000-lime-admin-common.scss [hunk 5]
```diff
@@ -4356,4 +4356,79 @@ html[dir="rtl"] th {
     #rel-eq-th, td.relevance-equation {
         display: none;
     }
+}
+
+/** Styles for Data Entry screen */
+#responsedetail {
+    .checkbox {
```
Comment (Shnoulle) [inline assets/packages/adminbasics/scss/000-lime-admin-common.scss:4363]: `unseen-item` or `unseen-checkbox` ?

File: application/controllers/admin/DataEntry.php [hunk 6]
```diff
@@ -1325,6 +1329,23 @@ public function editdata($subaction, $id, $surveyid)
...
+                    $aDataentryoutput .= "<div class=\"answers-list {$answerWrapperClass}\">";
+                    foreach ($questionInputs as $questionInputField => $questionInput) {
+                        $aDataentryoutput .= "<div class=\"answer-item\">";
+                        $aDataentryoutput .= "<div class=\"checkbox unseen-item\">"
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:1336]: `unseen-checkbox` for class ?

**Review by Shnoulle — APPROVED** [review]
Really great work ! And lot of work.

Just some details and confirmation

**Comment by Shnoulle** [comment]
I think it's best if unseen is after answers here.

Maybe 3 columns ? (with some cell with colspan)

**Comment by Shnoulle** [comment]
For need update by author

1. To confirm : debug=2 throw error
2. Single choice with comment don't seems to get the good unseen state

To test : import https://github.com/LimeSurvey/LimeSurvey/blob/3.x-LTS/docs/demosurveys/ls3_sample_survey_multilingual_fr_de_en_it.lss
activate, launch , move next : response is created with ALL unseen value

Capture d’écran du 2023-05-05 09-21-56

**Comment by Shnoulle** [comment]
Another update to do : for numeric and date : NULL value is set for empty, 
Then 2 solution (without updating DB)

1. Show _Unseen or not answered_
2. Don't show checkbox

To test : same survey, goes at end and submit

Capture d’écran du 2023-05-05 16-31-19
Capture d’écran du 2023-05-05 16-31-37

**Comment by gabrieljenik** [comment]
@Shnoulle Ready for your reviewal.

Added validation: Unseen can onlybe checked if answer is empty.
Had to add some options for making answers empty on some control types (this screen was rather outdated)
Had to add a hook mechanism on the save.js, as to run custom validation before submitting a form.
(was on 3.x, not on v5, nor v6)

**Comment by Shnoulle** [comment]
Still the issue with numeric + multi-numeric + + date

File: application/controllers/admin/DataEntry.php [hunk 7]
```diff
@@ -637,20 +648,25 @@ public function editdata($subaction, $id, $surveyid)
...
+                        $questionInputs[$fname['fieldname']] = "";
                         break;
                     case Question::QT_Q_MULTIPLE_SHORT_TEXT:
-                        $aDataentryoutput .= $fname['subquestion'] . '&nbsp;';
-                        $aDataentryoutput .= CHtml::textField($fname['fieldname'], $idrow[$fname['fieldname']]);
+                        $questionInput = $fname['subquestion'] . '&nbsp;';
+                        $questionInput .= CHtml::textField($fname['fieldname'], $idrow[$fname['fieldname']]);
+                        $questionInputs[$fname['fieldname']] = $questionInput;
                         break;
                     case Question::QT_K_MULTIPLE_NUMERICAL:
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:661]: No unsees state for QT_K_MULTIPLE_NUMERICAL

File: application/controllers/admin/DataEntry.php [hunk 8]
```diff
@@ -659,16 +675,24 @@ public function editdata($subaction, $id, $surveyid)
...
+                        }
+                        //Add 'No Answer'
+                        $questionInput .= "<input type='radio' class='' name='{$fname['fieldname']}' value=''";
+                        if ($idrow[$fname['fieldname']] == '') {
+                            $questionInput .= " checked";
                         }
+                        $questionInput .= " />" . gT("No answer") . "&nbsp;\n";
+                        $questionInputs[$fname['fieldname']] = $questionInput;
                         break;
                     case Question::QT_D_DATE: //DATE
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:697]: No unseen state for QT_D_DATE

File: application/controllers/admin/DataEntry.php [hunk 9]
```diff
@@ -902,103 +940,96 @@ public function editdata($subaction, $id, $surveyid)
...
+                                });
                             });
                             </script>';
-                        } else {
-                            //file count
-                            $aDataentryoutput .= '<input readonly id="' . $fname['fieldname'] . '" name="' . $fname['fieldname'] . '" value ="' . htmlspecialchars($idrow[$fname['fieldname']]) . '" /></td></table>';
-                        }
+                        $questionInputs[$baseFieldName] = $questionInput;
                         break;
                     case Question::QT_N_NUMERICAL: //NUMERICAL TEXT
```
Comment (Shnoulle) [inline application/controllers/admin/DataEntry.php:1034]: No unseen state for QT_N_NUMERICAL

**Review by Shnoulle — CHANGES_REQUESTED** [review]
Still the issue with numeric + multi-numeric + + date

**Review by Shnoulle — APPROVED** [review]
All done !

Except conflict :(

**Review by Shnoulle — CHANGES_REQUESTED** [review]
No 400 erro if JS broken : savec value and send flash messages

**Review by Shnoulle — APPROVED** [review]

**Pull request merged**